cirq/linalg/decompositions.py

```diff
@@ -854,3 +857,129 @@ def kak_decomposition(unitary_object: Union[np.ndarray, 'cirq.SupportsUnitary'],
+        raise ValueError(f'Expected input unitary to have shape (...,4,4), but'
+                         'got {unitary.shape}.')
+
+    if atol <= 0:
+        raise ValueError(f'Input atol must be positive, got {atol}.')
+
+    UB = np.einsum('...ab,...bc,...cd', MAGIC_CONJ_T, unitary, MAGIC)
+
+    m = np.einsum('...ab,...cb', UB, UB)
+
+    evals, _ = np.linalg.eig(m)
+
+    # The algorithm in the appendix mentioned above is slightly incorrect in
+    # that it only works for elements of SU(4). A phase correction must be
+    # added to deal with U(4).
+    phases = np.log(-1j * np.linalg.det(unitary)).imag + np.pi / 2
+    evals *= np.exp(-1j * phases / 2)[..., np.newaxis]
+
+    # The following steps follow the appendix exactly.
+    S2 = np.log(-1j * evals).imag + np.pi / 2
+    S2 = np.sort(S2, axis=-1)[..., ::-1]
+
+    n_shifted = (np.round(S2.sum(axis=-1) / (2 * np.pi))).astype(int)
+    for n in range(1, 5):
+        S2[n_shifted == n, :n] -= 2 * np.pi
+
+    # Fix pathological case of SWAP gate
+    S2[n_shifted == -1, :3] += 2 * np.pi
+
+    k_vec = (np.einsum('ab,...b', KAK_GAMMA, S2))[..., 1:] / 2
+
+    return _canonicalize_kak_vector(k_vec, atol)
+
+
+def _canonicalize_kak_vector(k_vec: np.ndarray, atol: float) -> np.ndarray:
+    r"""Map a KAK vector into its Weyl chamber equivalent vector.
+
+    This implementation is vectorized but does not produce the single qubit
+    unitaries required to bring the KAK vector into canonical form.
+
+    Args:
+        k_vec: THe KAK vector to be canonicalized. This input may be vectorized,
+            with shape (...,3), where the final axis denotes the k_vector and
+            all other axes are broadcast.
+        atol: How close x2 must be to π/4 to guarantee z2 >= 0.
+
+    Returns:
+        The canonicalized decomposition, with vector coefficients (x2, y2, z2)
+        satisfying:
+
+            0 ≤ abs(z2) ≤ y2 ≤ x2 ≤ π/4
+            if x2 = π/4, z2 >= 0
+        The output is vectorized, with shape k_vec.shape[:-1] + (3,).
+    """
+
+    # Get all strengths to (-¼π, ¼π]
+    k_vec = np.mod(k_vec + np.pi / 4, np.pi / 2) - np.pi / 4
+
+    # Sort in descending order with respect to absolute value.
+    order = np.argsort(np.abs(k_vec), axis=-1)
+    k_vec = np.take_along_axis(k_vec, order, axis=-1)[..., ::-1]
+
+    # Multiply x,z and y,z components by -1 to fix x,y sign.
+    x_negative = k_vec[..., 0] < 0
+    if np.any(x_negative):
+        k_vec[x_negative, 0] *= -1
+        k_vec[x_negative, 2] *= -1
+    y_negative = k_vec[..., 1] < 0
+    if np.any(y_negative):
+        k_vec[y_negative, 1] *= -1
+        k_vec[y_negative, 2] *= -1
+
+    # If x = π/4, force z to be positive.
+    x_is_pi_over_4 = np.isclose(k_vec[..., 0], np.pi / 4, atol=atol)
+    z_is_negative = k_vec[..., 2] < 0
+    need_diff = np.logical_and(x_is_pi_over_4, z_is_negative)
+    if np.any(need_diff):  # -1 to x and z components, then shift x up by pi/4
```

Comment:
The comment differs from what the code does.

Reply:
I added a clarification. The comment refers to the operations allowed under local unitaries.